File: backend/app/api/ingest.py
"""Ingestion API: a real, live connector for workflow step 1.

Turns a pasted or uploaded interaction (meeting note, transcript, email, CRM
rows, support ticket) into citeable, retrievable evidence so the next decision
run grounds on it immediately. Fully offline-safe: the in-memory corpus is always
updated; pgvector persistence is added transparently when a database is set.

Every ingest is org-scoped (the caller's org owns the evidence) and runs a light,
deterministic signal-extraction pass (no LLM): it returns a short title plus a
few candidate signal phrases so the interaction surfaces structured hints the
moment it lands. Repeat pastes are idempotent per org.

Endpoints:
  * ``POST /ingest``         -> ingest raw text, returns id + extracted signals.
  * ``GET  /ingest/sources`` -> recognized source types for the UI select.
  * ``GET  /ingest/recent``  -> the org's recently ingested interactions.
  * ``POST /ingest/web``     -> optional live web-search ingest (best-effort).
"""

from __future__ import annotations

import logging
import time
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.api._ingest_signals import (
    content_fingerprint,
    derive_title,
    extract_signals,
    recent_store,
)
from app.api._org import current_org
from app.retrieval.connectors import (
    TextImportConnector,
    WebSearchConnector,
    recognized_sources,
)

logger = logging.getLogger("app.api.ingest")

router = APIRouter(tags=["ingest"])

_text_connector = TextImportConnector()
_web_connector = WebSearchConnector()


def _record_from_result(
    result_dict: Dict[str, Any],
    *,
    title: str,
    signals: List[Dict[str, str]],
    chars: int,
) -> Dict[str, Any]:
    """Project a connector result + extracted signals into the API response.

    Preserves the existing connector fields (``chunks_written``, ``ids``,
    ``persisted`` ...) and adds the ingest-experience fields (``id``,
    ``extracted_signals``, ``chars``, ``title``) the UI surfaces.
    """
    enriched = dict(result_dict)
    enriched["id"] = result_dict.get("doc_id") or ""
    enriched["title"] = title
    enriched["extracted_signals"] = signals
    enriched["chars"] = chars
    enriched["ts"] = enriched.get("ts") or time.time()
    return enriched


# ---------------------------------------------------------------------------
# Request / response models
# ---------------------------------------------------------------------------
class IngestIn(BaseModel):
    """Raw interaction text to ingest into the retrieval corpus."""

    text: str = Field(description="The raw interaction text (note, transcript, email, CSV rows).")
    source_type: str = Field(default="document", description="Recognized source type key.")
    title: Optional[str] = Field(default=None, description="Optional human title for the import.")
    account_id: Optional[str] = Field(default=None, description="Optional account to scope the evidence to.")
    domain: str = Field(default="customer_success", description="Domain pack the corpus belongs to.")


class WebIngestIn(BaseModel):
    """A live web-search query to fetch and ingest as web evidence."""

    query: str = Field(description="Search query to fetch public context for.")
    account_id: Optional[str] = Field(default=None, description="Optional account to scope to.")
    title: Optional[str] = Field(default=None, description="Optional title override.")
    domain: str = Field(default="customer_success", description="Domain pack the corpus belongs to.")


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------
@router.get("/ingest/sources")
async def list_sources() -> Dict[str, List[Dict[str, str]]]:
    """Return the recognized interaction source types (for the UI select)."""
    return {"sources": recognized_sources()}


@router.get("/ingest/recent")
async def list_recent(
    account_id: Optional[str] = None,
    limit: int = 25,
    org_id: str = Depends(current_org),
) -> Dict[str, List[Dict[str, Any]]]:
    """Return the caller's recently ingested interactions, most recent first.

    Optionally filter by ``account_id``. Org-scoped: an org only ever sees the
    evidence it ingested this process.
    """
    return {"items": recent_store.list(org_id, account_id=account_id, limit=limit)}


@router.post("/ingest")
async def ingest(
    body: IngestIn, org_id: str = Depends(current_org)
) -> Dict[str, Any]:
    """Ingest raw interaction text and make it retrievable in the next run.

    Org-scoped and idempotent: re-pasting the same content for the same account
    returns the original record instead of duplicating evidence. Runs a light,
    deterministic signal-extraction pass so the response carries a short title
    and a few candidate signal phrases (no LLM, fully offline).
    """
    if not (body.text or "").strip():
        raise HTTPException(status_code=400, detail="text must not be empty")

    title = derive_title(body.text, body.source_type, body.title)
    signals = extract_signals(body.text)
    chars = len(body.text.strip())

    # Idempotency: a repeat paste for the same org + account + type short-circuits
    # to the stored record so we never double-index the same evidence.
    fingerprint = content_fingerprint(org_id, body.account_id, body.source_type, body.text)
    existing = recent_store.get(org_id, fingerprint)
    if existing is not None:
        return {**existing, "duplicate": True}

    try:
        result = await _text_connector.ingest(
            text=body.text,
            source_type=body.source_type,
            title=title,
            account_id=body.account_id,
            domain=body.domain,
            org_id=org_id,
        )
    except Exception as exc:  # noqa: BLE001 - surface a clean error, never 500-trace
        logger.exception("ingest failed")
        raise HTTPException(status_code=500, detail=f"ingest failed: {exc}") from exc

    if not result.ok:
        raise HTTPException(status_code=422, detail=result.detail or "nothing ingested")

    record = _record_from_result(
        result.to_dict(), title=title, signals=signals, chars=chars
    )
    recent_store.add(org_id, fingerprint, record)
    return {**record, "duplicate": False}


@router.post("/ingest/web")
async def ingest_web(
    body: WebIngestIn, org_id: str = Depends(current_org)
) -> Dict[str, Any]:
    """Optional: run a live web search and ingest its snippets (best-effort).

    Returns ``ok: false`` with a reason when the network is unavailable or the
    query has no public results, so the UI can degrade without an error toast.
    """
    if not (body.query or "").strip():
        raise HTTPException(status_code=400, detail="query must not be empty")

    try:
        result = await _web_connector.ingest(
            query=body.query,
            account_id=body.account_id,
            title=body.title,
            domain=body.domain,
            org_id=org_id,
        )
    except Exception as exc:  # noqa: BLE001 - never hard-fail the optional path
        logger.warning("web ingest failed: %s", exc)
        return {
            "ok": False,
            "chunks_written": 0,
            "ids": [],
            "source_type": "web",
            "domain": body.domain,
            "extracted_signals": [],
            "detail": "web search unavailable",
        }

    result_dict = result.to_dict()
    if not result_dict.get("ok"):
        return result_dict

    title = derive_title(
        result_dict.get("title") or body.query, "web", body.title or body.query
    )
    record = _record_from_result(
        result_dict, title=title, signals=[], chars=result_dict.get("chunks_written", 0)
    )
    fingerprint = content_fingerprint(org_id, body.account_id, "web", body.query)
    recent_store.add(org_id, fingerprint, record)
    return record

File: frontend/lib/api/ingest.ts
// Typed client for the live ingestion connector (workflow step 1).
//
// POST /ingest turns pasted or uploaded interaction text into citeable,
// retrievable evidence. GET /ingest/sources lists the recognized source types.
// POST /ingest/web runs an optional best-effort live web search.
//
// Reads (sources) fall back to a local list so the select renders even if the
// backend is briefly unreachable. The import mutation surfaces real errors so
// the user knows whether their evidence actually landed.

import { API_BASE } from "@/lib/api";
import { authHeaders } from "@/lib/auth";

export interface IngestSource {
  key: string;
  label: string;
  description: string;
}

export interface IngestRequest {
  text: string;
  source_type: string;
  title?: string | null;
  account_id?: string | null;
  domain?: string;
}

// A deterministically extracted candidate signal (offline, no LLM): a sentence
// that mentions churn / renewal / usage / sentiment, tagged with its category.
export interface ExtractedSignal {
  category: string;
  label: string;
  keyword: string;
  text: string;
}

export interface IngestResponse {
  ok: boolean;
  chunks_written: number;
  ids: string[];
  doc_id: string;
  id: string;
  account_id: string | null;
  source_type: string;
  title: string;
  domain: string;
  embed_method: string;
  persisted: string;
  detail: string;
  extracted_signals: ExtractedSignal[];
  chars: number;
  duplicate?: boolean;
  ts?: number;
}

export interface WebIngestRequest {
  query: string;
  account_id?: string | null;
  title?: string | null;
  domain?: string;
}

// Mirrors backend SOURCE_TYPES so the select is populated even offline.
export const DEFAULT_SOURCES: IngestSource[] = [
  { key: "meeting_notes", label: "Meeting notes", description: "Notes from a call, QBR, or internal sync." },
  { key: "call_transcript", label: "Call transcript", description: "Verbatim transcript of a customer or sales call." },
  { key: "email", label: "Email", description: "An inbound or outbound email thread." },
  { key: "support_ticket", label: "Support ticket", description: "A support or success ticket and its conversation." },
  { key: "crm_record", label: "CRM record", description: "Pasted CRM fields, account notes, or exported rows." },
  { key: "chat_message", label: "Chat message", description: "Slack, Teams, or in-app chat exchange." },
  { key: "document", label: "Document", description: "A general document, brief, or knowledge article." },
  { key: "web", label: "Web result", description: "Snippet captured from a live web search." },
];

// Realistic example text per source type, so a user can try the experience in
// one click. Keyed by source key; falls back to EXAMPLE_FALLBACK.
export const EXAMPLE_FALLBACK =
  "QBR with the champion. Adoption of the analytics module stalled, two power users left, and they are evaluating a competitor before renewal. The economic buyer is concerned about pricing.";

export const SOURCE_EXAMPLES: Record<string, string> = {
  meeting_notes:
    "QBR with Zephyr Dynamics. Adoption of the analytics module stalled after the Q1 reorg, and two power users left. The champion is happy with support but the economic buyer is concerned about pricing ahead of the May renewal. They mentioned evaluating a competitor.",
  call_transcript:
    "CSM: How has the rollout gone since launch?\nCustomer: Honestly, usage has dropped. Two of our admins left and onboarding the new team stalled. We are frustrated and are looking at alternatives before the renewal in March.",
  email:
    "Subject: Renewal and open issues\n\nHi team, ahead of our renewal we wanted to flag that adoption has been lower than expected and we have an open support ticket about slow dashboards. Leadership is asking about ROI. Can we discuss pricing options?",
  support_ticket:
    "Ticket #4821 (high priority): Dashboards time out for users on the Growth plan. Customer reports this has happened three times this week and is frustrated. Renewal is in 30 days and they are escalating.",
  crm_record:
    "Account: Zephyr Dynamics. Stage: At risk. ARR: 84000. Health: 41. Renewal: 2026-05-15. Last touch: champion flagged stalled adoption and a competitor evaluation. Owner: Priya.",
  chat_message:
    "[Slack #cs-zephyr] Champion just pinged: their VP is questioning the spend and usage is down this quarter. They want a business review before they commit to renewal. Flagging churn risk.",
  document:
    "Internal brief: Zephyr Dynamics expansion blocked by low adoption. Two power users left; onboarding for the replacement admins has not started. Competitor evaluation underway. Renewal at risk in May.",
  web: "",
};

export function exampleForSource(key: string): string {
  const ex = SOURCE_EXAMPLES[key];
  return ex !== undefined && ex !== "" ? ex : EXAMPLE_FALLBACK;
}

export interface RecentIngest {
  id: string;
  title: string;
  source_type: string;
  account_id: string | null;
  chars: number;
  chunks_written: number;
  persisted: string;
  extracted_signals: ExtractedSignal[];
  ts?: number;
}

/** The org's recently ingested interactions, most recent first. */
export async function getRecentIngests(
  signal?: AbortSignal,
): Promise<RecentIngest[]> {
  try {
    const res = await fetch(`${API_BASE}/ingest/recent`, {
      headers: authHeaders({ Accept: "application/json" }),
      credentials: "include",
      cache: "no-store",
      signal,
    });
    if (!res.ok) throw new Error(`GET /ingest/recent failed (${res.status})`);
    const data = (await res.json()) as { items?: RecentIngest[] };
    return Array.isArray(data.items) ? data.items : [];
  } catch {
    return [];
  }
}

export async function getIngestSources(
  signal?: AbortSignal,
): Promise<IngestSource[]> {
  try {
    const res = await fetch(`${API_BASE}/ingest/sources`, {
      headers: authHeaders({ Accept: "application/json" }),
      credentials: "include",
      cache: "no-store",
      signal,
    });
    if (!res.ok) throw new Error(`GET /ingest/sources failed (${res.status})`);
    const data = (await res.json()) as { sources?: IngestSource[] };
    return data.sources && data.sources.length ? data.sources : DEFAULT_SOURCES;
  } catch {
    return DEFAULT_SOURCES;
  }
}

/** Ingest raw interaction text. Throws on failure so the UI can report it. */
export async function ingestText(
  payload: IngestRequest,
): Promise<IngestResponse> {
  const res = await fetch(`${API_BASE}/ingest`, {
    method: "POST",
    headers: authHeaders({ "Content-Type": "application/json" }),
    credentials: "include",
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    let detail = `Import failed (${res.status})`;
    try {
      const body = (await res.json()) as { detail?: string };
      if (body.detail) detail = body.detail;
    } catch {
      /* keep the status-based message */
    }
    throw new Error(detail);
  }
  return (await res.json()) as IngestResponse;
}

/**
 * Optional live web-search ingest. Best-effort: returns ok:false (rather than
 * throwing) when the network is unavailable or there are no public results.
 */
export async function ingestWeb(
  payload: WebIngestRequest,
): Promise<IngestResponse> {
  try {
    const res = await fetch(`${API_BASE}/ingest/web`, {
      method: "POST",
      headers: authHeaders({ "Content-Type": "application/json" }),
      credentials: "include",
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      return {
        ok: false,
        chunks_written: 0,
        ids: [],
        doc_id: "",
        id: "",
        account_id: payload.account_id ?? null,
        source_type: "web",
        title: payload.title ?? "",
        domain: payload.domain ?? "customer_success",
        embed_method: "none",
        persisted: "none",
        extracted_signals: [],
        chars: 0,
        detail: `Web search failed (${res.status})`,
      };
    }
    return (await res.json()) as IngestResponse;
  } catch {
    return {
      ok: false,
      chunks_written: 0,
      ids: [],
      doc_id: "",
      id: "",
      account_id: payload.account_id ?? null,
      source_type: "web",
      title: payload.title ?? "",
      domain: payload.domain ?? "customer_success",
      embed_method: "none",
      persisted: "none",
      extracted_signals: [],
      chars: 0,
      detail: "Web search unavailable (offline).",
    };
  }
}
